ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/util/PDEXmlProcessorFactory.java

```diff
@@ -0,0 +1,166 @@
+ *
+ * @see <a href="https://rules.sonarsource.com/java/RSPEC-2755/">RSPEC-2755</a>
+ */
+public class PDEXmlProcessorFactory {
+	private PDEXmlProcessorFactory() {
+		// static Utility only
+	}
+
+	private static final DocumentBuilderFactory DOCUMENT_BUILDER_FACTORY_ERROR_ON_DOCTYPE = createDocumentBuilderFactoryWithErrorOnDOCTYPE();
+	private static final SAXParserFactory SAX_FACTORY_ERROR_ON_DOCTYPE = createSAXFactoryWithErrorOnDOCTYPE(false);
+	private static final SAXParserFactory SAX_FACTORY_ERROR_ON_DOCTYPE_NS = createSAXFactoryWithErrorOnDOCTYPE(true);
+	private static final SAXParserFactory SAX_FACTORY_IGNORING_DOCTYPE = createSAXFactoryIgnoringDOCTYPE();
+
+	/**
+	 * Creates TransformerFactory which throws TransformerException when
+	 * detecting external entities.
+	 *
+	 * @return javax.xml.transform.TransformerFactory
+	 */
+	public static TransformerFactory createTransformerFactoryWithErrorOnDOCTYPE() {
+		TransformerFactory factory = TransformerFactory.newInstance();
+		// prohibit the use of all protocols by external entities:
+		factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, ""); //$NON-NLS-1$
+		factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, ""); //$NON-NLS-1$
+		return factory;
+	}
+
+	/**
+	 * Creates DocumentBuilderFactory which throws SAXParseException when
+	 * detecting external entities. It's magnitudes faster to call
+	 * {@link #createDocumentBuilderWithErrorOnDOCTYPE()}.
+	 *
+	 * @return javax.xml.parsers.DocumentBuilderFactory
+	 */
+	public static synchronized DocumentBuilderFactory createDocumentBuilderFactoryWithErrorOnDOCTYPE() {
+		DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+		// completely disable DOCTYPE declaration:
+		try {
+			factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true); //$NON-NLS-1$
+		} catch (ParserConfigurationException e) {
+			throw new RuntimeException(e.getMessage(), e);
+		}
+		return factory;
+	}
+
+	/**
+	 * Creates DocumentBuilder which throws SAXParseException when detecting
+	 * external entities. The builder is not thread safe.
+	 *
+	 * @return javax.xml.parsers.DocumentBuilder
+	 * @throws ParserConfigurationException
+	 */
+	public static DocumentBuilder createDocumentBuilderWithErrorOnDOCTYPE() throws ParserConfigurationException {
+		return DOCUMENT_BUILDER_FACTORY_ERROR_ON_DOCTYPE.newDocumentBuilder();
+	}
+
+	/**
+	 * Creates DocumentBuilderFactory which throws SAXParseException when
+	 * detecting external entities.
+	 *
+	 * @return javax.xml.parsers.DocumentBuilderFactory
+	 */
+	public static SAXParserFactory createSAXFactoryWithErrorOnDOCTYPE() {
+		SAXParserFactory f = SAXParserFactory.newInstance();
+		try {
+			// force org.xml.sax.SAXParseException for any DOCTYPE:
+			f.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true); //$NON-NLS-1$
+		} catch (Exception e) {
+			throw new RuntimeException(e);
+		}
+		return f;
+	}
+
```

Comment:
I would like to keep it as a fall back if we discover a case where an exception is too harsh.

Reply:
For which one?
We should not keep internal methods for other projects, because if it is not used it can happen that one day in the future somebody comes along and cleans up this class and then that other project suddenly breaks. Everything that is not API can be removed at any time. Furthermore even a method that is not called has maintanance cost, because somebody reading this class has to understand what it is doing and then wonders who's calling it and why no code in PDE does that. In that regard no code is the best code.


And at least for the Eclipse SDK (Equinox, Platform, JDT and PDE), PDE is at the end of the dependency hierarchy, so other projects are outside the SDK? Or, since you mentioned copy-pasting, do you want to add this class to other Eclipse SDK projects as it is?
Then I suggest to add it to one at the top of the hierarchy, probably equinox or eclipse(-runtime), as an internal class, add a comment that states that this is intended to be used in the entire Eclipse-SDK (Equinox, Platform, JDT and PDE, but not more) so that one knows where to search callers.
This way the code can be re-used, but we don't have to make it an API.

Reply:
I struggled to find a place that both needs this methods itself (which would make sure it is not deleted because unused and not API) and is available in all other modules as well. I would highly appreciate help to find a good place, Otherwise i would copy & paste the (small) solution to other projects and keep only the test at one module.
Part of it would be needed for example at 
equinox ExtensionRegistry or org.eclipse.core.internal.resources but in at least 8 distinct git repositories. I think it's very bad to share no-API across repositories.
@jonahgraham any idea?

Reply:
We could also register a provider of such factories as (declarative) OSGi Service and can then query the service from any bundle in the runtime. Registering it in something like equinox.common. maybe even cascade it.

But I'm fine doing that in a later step and for now submit this as it is.